compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs

```diff
@@ -1119,6 +1120,16 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                         Applicability::MachineApplicable,
                     );
                     return true;
+                } else if let Some(expected_ty_expr) = expected_ty_expr {
```

Comment:
Shouldn't this actually check that `.as_ref()` actually returns the right type?

Reply:
We check before invoking this that the type we have is either a `Result` or `Option`, and before suggesting anything check that the `Some`/`Ok` type is `&T`, where `T` `can_eq` with our expected type. As far as I can _see_ this should guarentee that `as_ref` returns `&T`.